File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        // Now the logic is that if the first and the last elements are adjecent, then the simple logic says that only one of them can be in the answer. So used the normal logic of House Robber (1st part of this question) for both first excluding the first element and then excluding the last element.
        int n = nums.length;
        int prev1 = nums[0];
        int prev2 = 0;
        for(int i=1; i<n-1; i++) {
            int pick = nums[i];
            if(i > 1) pick += prev2;
            int nonPick = prev1;

            int curr = Math.max(pick, nonPick);
            prev2 = prev1;
            prev1 = curr;
        }
        int firstAns = prev1;

        if(n > 1) prev1 = nums[1];
        prev2 = 0;
        for(int i=2; i<n; i++) {
            int pick = nums[i];
            if(i > 2) pick += prev2;
            int nonPick = prev1;

            int curr = Math.max(pick, nonPick);
            prev2 = prev1;
            prev1 = curr;
        }
        int lastAns = prev1;

        return Math.max(firstAns, lastAns);
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if(n == 1):
            return nums[0]
        dp = [nums[0]]

        for i in range(1, n-1):
            pick = nums[i]
            if(i > 1):
                pick += dp[i-2]
            notPick = dp[i-1]

            dp.append(max(pick, notPick))

        max1 = dp[n-2]

        dp = [nums[1]]

        for i in range(2, n):
            pick = nums[i]
            if(i > 2):
                pick += dp[i-3]
            notPick = dp[i-2]

            dp.append(max(pick, notPick))

        max2 = dp[n-2]

        return max(max1, max2)